subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -333,13 +333,47 @@ public NodeInVisitingSegment apply(TaskNode taskNode) {
     }
 
     @Override
-    public Set<Task> getDependencies(Task task) {
-        TaskNode node = nodeMapping.get(task);
-        ImmutableSet.Builder<Task> builder = ImmutableSet.builder();
-        for (Node dependencyNode : node.getDependencySuccessors()) {
-            dependencyNode.collectTaskInto(builder);
-        }
-        return builder.build();
+    public ExecutionDependencies getDependencies(Task task) {
+        return getDependencies(nodeMapping.get(task));
+    }
+
+    @Override
+    public ExecutionDependencies getDependencies(TransformationNodeIdentifier transformation) {
+        return getDependencies(nodeMapping.get(transformation));
+    }
```

Comment:
There should be only one method here that takes a `WorkIdentifier`. It should also use a visitor instead of constructing expensive collections.